homeassistant/components/omnilogic/sensor.py

```diff
@@ -0,0 +1,688 @@
+            coordinator=coordinator,
+            kind=kind,
+            name=name,
+            backyard=backyard,
+            bow=bow,
+            icon=icon,
+            entitydata=sensordata,
+        )
+        self._state = None
+        self._unit_type = backyard["Unit-of-Measurement"]
+        self._device_class = device_class
+        self._unit = None
+
+    @property
+    def device_class(self):
+        """Return the device class of the entity."""
+        if self._device_class != "none":
+            return self._device_class
+
+    @property
+    def unit_of_measurement(self):
+        """Return the right unit of measure."""
+        return self._unit
+
+    @property
+    def force_update(self):
+        """Force update."""
+        return True
+
+    def add_alarms(self, alarms: dict):
+        """Add Alarm attributes."""
+        self.alarms = alarms
+
+        self._attrs["Alarm"] = ""
+        if len(self.alarms) != 0:
+            self._attrs["Alarm"] = (
+                self.alarms[0]["Message"] + " (" + self.alarms[0]["Comment"] + ")"
+            )
+
+    @staticmethod
+    def find_chlorinator(coordinator, systemid):
+        """Find the correct chlorinator entity."""
+
+        for backyard in coordinator.data:
+            for bow in backyard.get("BOWS"):
+                if bow["Chlorinator"]["systemId"] == systemid:
+                    sensordata = bow.get("Chlorinator")
+                    break
+
+        return sensordata
+
+    @staticmethod
+    def find_csad(coordinator, systemid):
+        """Find the correct CSAD entity."""
+
+        for backyard in coordinator.data:
+            for bow in backyard.get("BOWS"):
+                if bow["CSAD"]["systemId"] == systemid:
+                    sensordata = bow.get("CSAD")
+                    break
+
+        return sensordata
+
+
+class OmniLogicTemperatureSensor(OmnilogicSensor):
+    """Define an OmniLogic Temperature (Air/Water) Sensor."""
+
+    def __init__(
+        self,
+        coordinator: OmniLogicUpdateCoordinator,
+        kind: str,
+        name: str,
+        backyard: dict,
+        bow: dict,
+        device_class: str,
+        icon: str,
+        unit: str,
+        sensordata: dict,
+    ):
+        """Initialize the sensor."""
+        super().__init__(
+            coordinator=coordinator,
+            kind=kind,
+            name=name,
+            backyard=backyard,
+            bow=bow,
+            device_class=device_class,
+            icon=icon,
+            unit=unit,
+            sensordata=sensordata,
+        )
+
+    @property
+    def state(self):
+        """Return the state for the temperature sensor."""
+        _LOGGER.debug("Updating state of sensor: %s", self._name)
+
+        sensordata = None
+
+        if self._kind == "water_temperature":
+            for backyard in self.coordinator.data:
+                for bow in backyard["BOWS"]:
+                    if bow["systemId"] == self.bow["systemId"]:
```

Comment:
Is the system id unique over all backyards? In that case we could pivot the data structure and store the data per system id instead of backyard. Then under each system id, we store a map between kind and data value.

If we can make this pivot as early as possible we can make it clearer and cleaner downstream, ie here where we need to look up the data.

Reply:
The system id is unique per back yard.

Reply:
Good! Then I'd try to remake the data structure per above in the coordinator update method before returning the data.

Reply:
Ah - actually misread what you were looking at here. The structure is:

for each user we may receive:

one or more backyard = unique system_id for each omnilogic system configured on the account.
-> one or more bows (body of water / pool) which have their own unique ids

within the backyard object there may be one or more control items (pumps, relays, etc)

within each pool object there may be one or more control items (filters, pumps, relays, etc)

So we can't pivot to use the bow system id at the coordinator level easily, there is too much potential variation in system configurations below the backyard.

Reply:
Does each item have its own unique id and is that unique across the whole platform? Can you upload an example of the data?

Reply:
Unfortunately not.

For example in a multi-pool setup:

```
backyard(system)_id = 123456
  lights_relay1 (landscape lights) system_id = 3
  sprinkler_relay1 (sprinkler system) system_id = 4
  bow1_id (pool) = 1
    filter_pump system_id = 5
    heater system_id = 6
    feature_pump system_id = 7
    lights system_id = 8
  bow2_id (spa) = 2
    pump system_id = 9
    lights system_id = 10
backyard(system)_id = 654321
  lights_relay1 system_id = 3
  bow1_id (pool) system_id = 2
    filter_pump system_id = 4
    chlorinator system_id = 5
    lights system_id = 6
```
The data is also dynamic and unique to each pool - so if you added a new relay to your pool for a new water feature or added a new relay to the system it _may_ bump the system IDs out of whack from what was saved which is why we have to parse each time to find the correct item. IDs may be duplicated between backyards, and based on the documentation may also be duplicated within a backyard in separate bodies of water (bow).

Reply:
How do we handle changing ids with the current solution? I don't see that.

Reply:
It is only handled on a restart of the system or by removing / adding the integration. The unique id doesn't use the system id since it might change, but we didn't handle dynamic changing while running since it is a remote edge case that we haven't yet seen in any of our test setups. Unfortunately we still have the duplication issue within pools so we had to use this approach to identifying the correct elements in the updates.

Reply:
Would it work to use an id like this, where the words in capital letters are the actual ids:
```py
global_id = (("backyard", BACKYARD_ID), ("bow", BOW_ID), ("pump", PUMP_ID))
```

Reply:
Yes - that would be similar to the unique_id we have stored currently (however we used the names from Omnilogic in that case). So then the suggestion here is to rewrite the coordinator update to parse the full data and create a new data structure which could be addressed through this pseudo-id?

Reply:
To help understand, here is a sample data structure from one of the more complex setups we are testing with:

```
[
   {
      "systemId":"24353",
      "statusVersion":"5",
      "airTemp":"105",
      "status":"1",
      "state":"1",
      "configUpdatedTime":"0001-01-01T00:00:00",
      "datetime":"2020-08-25T12:02:31.5773084",
      "Relays":[
         {
            "systemId":"17",
            "relayState":"1",
            "Name":"Landscaping",
            "Type":"RLY_HIGH_VOLTAGE_RELAY",
            "Function":"RLY_POOL_LIGHT",
            "Alarms":[

            ]
         }
      ],
      "BOWS":[
         {
            "systemId":"1",
            "flow":"255",
            "waterTemp":"92",
            "Name":"Pool",
            "Supports-Spillover":"no",
            "Filter":{
               "systemId":"2",
               "valvePosition":"1",
               "filterSpeed":"90",
               "filterState":"1",
               "lastSpeed":"50",
               "Name":"Filter Pump",
               "Shared-Type":"BOW_NO_EQUIPMENT_SHARED",
               "Filter-Type":"FMT_VARIABLE_SPEED_PUMP",
               "Max-Pump-Speed":"100",
               "Min-Pump-Speed":"18",
               "Max-Pump-RPM":"3450",
               "Min-Pump-RPM":"600",
               "Priming-Enabled":"yes",
               "Alarms":[

               ]
            },
            "VirtualHeater":{
               "systemId":"15",
               "Current-Set-Point":"90",
               "enable":"no"
            },
            "Heater":{
               "systemId":"16",
               "heaterState":"0",
               "enable":"yes",
               "Shared-Type":"BOW_NO_EQUIPMENT_SHARED",
               "Operation":{
                  "VirtualHeater":{
                     "System-Id":"16",
                     "Name":"Heat Pump",
                     "Type":"PET_HEATER",
                     "Heater-Type":"HTR_HEAT_PUMP",
                     "Enabled":"yes",
                     "Priority":"HTR_PRIORITY_2",
                     "Run-For-Priority":"HTR_MAINTAINS_PRIORITY_FOR_AS_LONG_AS_VALID",
                     "Shared-Equipment-System-ID":"-1",
                     "Current-Set-Point":"90",
                     "Max-Water-Temp":"104",
                     "Min-Settable-Water-Temp":"65",
                     "Max-Settable-Water-Temp":"104",
                     "enable":"yes",
                     "systemId":"15"
                  }
               },
               "Alarms":[

               ]
            },
            "CSAD":{
               "systemId":"0",
               "ph":"",
               "orp":"",
               "status":"0",
               "mode":"0",
               "Alarms":[

               ]
            },
            "Group":{
               "systemId":"34",
               "groupState":"0"
            },
            "Lights":[
               {
                  "systemId":"11",
                  "lightState":"0",
                  "currentShow":"8",
                  "Name":"Wtfall Light",
                  "Type":"COLOR_LOGIC_UCL",
                  "Alarms":[

                  ]
               },
               {
                  "systemId":"25",
                  "lightState":"0",
                  "currentShow":"2",
                  "Name":"Main Light",
                  "Type":"COLOR_LOGIC_4_0",
                  "Alarms":[

                  ]
               }
            ],
            "Relays":[
               {
                  "systemId":"5",
                  "relayState":"1",
                  "Name":"Ozone",
                  "Type":"RLY_HIGH_VOLTAGE_RELAY",
                  "Function":"RLY_ACCESSORY",
                  "Alarms":[

                  ]
               },
               {
                  "systemId":"14",
                  "relayState":"0",
                  "Name":"Mister",
                  "Type":"RLY_HIGH_VOLTAGE_RELAY",
                  "Function":"RLY_ACCESSORY",
                  "Alarms":[

                  ]
               },
               {
                  "systemId":"26",
                  "relayState":"0",
                  "Name":"Water Ft",
                  "Type":"RLY_VALVE_ACTUATOR",
                  "Function":"RLY_WATER_FEATURE",
                  "Alarms":[

                  ]
               }
            ],
            "Pumps":[

            ]
         }
      ],
      "BackyardName":"Ponderosa",
      "Msp-Vsp-Speed-Format":"Percent",
      "Msp-Time-Format":"12 Hour Format",
      "Units":"Standard",
      "Msp-Chlor-Display":"Salt",
      "Msp-Language":"English",
      "Unit-of-Measurement":"Standard",
      "Alarms":[

      ],
      "Unit-of-Temperature":"UNITS_FAHRENHEIT"
   },
   {
      "systemId":"36211",
      "statusVersion":"5",
      "airTemp":"113",
      "status":"1",
      "state":"1",
      "configUpdatedTime":"0001-01-01T00:00:00",
      "datetime":"2020-08-25T12:15:09.4311922",
      "Relays":[
         {
            "systemId":"14",
            "relayState":"1",
            "Name":"Landscape",
            "Type":"RLY_HIGH_VOLTAGE_RELAY",
            "Function":"RLY_POOL_LIGHT",
            "Alarms":[

            ]
         },
         {
            "systemId":"29",
            "relayState":"0",
            "Name":"Misters",
            "Type":"RLY_HIGH_VOLTAGE_RELAY",
            "Function":"RLY_ACCESSORY",
            "Alarms":[

            ]
         }
      ],
      "BOWS":[
         {
            "systemId":"2",
            "flow":"138",
            "waterTemp":"96",
            "Name":"Pool",
            "Supports-Spillover":"no",
            "Filter":{
               "systemId":"3",
               "valvePosition":"1",
               "filterSpeed":"90",
               "filterState":"1",
               "lastSpeed":"100",
               "Name":"Filter Pump",
               "Shared-Type":"BOW_NO_EQUIPMENT_SHARED",
               "Filter-Type":"FMT_VARIABLE_SPEED_PUMP",
               "Max-Pump-Speed":"100",
               "Min-Pump-Speed":"22",
               "Max-Pump-RPM":"2800",
               "Min-Pump-RPM":"600",
               "Priming-Enabled":"yes",
               "Alarms":[

               ]
            },
            "VirtualHeater":{
               "systemId":"4",
               "Current-Set-Point":"90",
               "enable":"no"
            },
            "Heater":{
               "systemId":"5",
               "heaterState":"0",
               "enable":"yes",
               "Shared-Type":"BOW_NO_EQUIPMENT_SHARED",
               "Operation":{
                  "VirtualHeater":{
                     "System-Id":"5",
                     "Name":"Heat Pump",
                     "Type":"PET_HEATER",
                     "Heater-Type":"HTR_HEAT_PUMP",
                     "Enabled":"no",
                     "Priority":"HTR_PRIORITY_1",
                     "Run-For-Priority":"HTR_MAINTAINS_PRIORITY_FOR_AS_LONG_AS_VALID",
                     "Shared-Equipment-System-ID":"-1",
                     "Current-Set-Point":"90",
                     "Max-Water-Temp":"104",
                     "Min-Settable-Water-Temp":"65",
                     "Max-Settable-Water-Temp":"104",
                     "enable":"no",
                     "systemId":"4"
                  }
               },
               "Alarms":[

               ]
            },
            "CSAD":{
               "systemId":"0",
               "ph":"",
               "orp":"",
               "status":"0",
               "mode":"0",
               "Alarms":[

               ]
            },
            "Lights":[
               {
                  "systemId":"23",
                  "lightState":"0",
                  "currentShow":"0",
                  "Name":"Pool Light",
                  "Type":"COLOR_LOGIC_UCL",
                  "Alarms":[

                  ]
               },
               {
                  "systemId":"24",
                  "lightState":"0",
                  "currentShow":"0",
                  "Name":"Accent Light",
                  "Type":"COLOR_LOGIC_UCL",
                  "Alarms":[

                  ]
               }
            ],
            "Relays":[
               {
                  "systemId":"7",
                  "relayState":"0",
                  "Name":"Water Ft",
                  "Type":"RLY_VALVE_ACTUATOR",
                  "Function":"RLY_WATER_FEATURE",
                  "Alarms":[

                  ]
               }
            ],
            "Pumps":[

            ]
         }
      ],
      "BackyardName":"Indigo2",
      "Msp-Vsp-Speed-Format":"Percent",
      "Msp-Time-Format":"12 Hour Format",
      "Units":"Standard",
      "Msp-Chlor-Display":"Salt",
      "Msp-Language":"English",
      "Unit-of-Measurement":"Standard",
      "Alarms":[

      ],
      "Unit-of-Temperature":"UNITS_FAHRENHEIT"
   }
]
```

Reply:
Thanks for the data! Here's how I would try to do it. It may be incomplete since I don't know all the details. `fixture` is the json data you posted above:
```py
In [29]: item_kinds = ("BOWS", "Filter", "VirtualHeater", "Heater", "CSAD", "Group", "Lights", "Relays", "Pumps")

In [30]: data = {}

In [31]: def get_item_data(item, item_kind, current_id, item_kinds, data):
    ...:     if isinstance(item, list):
    ...:         for single_item in item:
    ...:             data = get_item_data(single_item, item_kind, current_id, item_kinds, data)
    ...:     if "systemId" in item:
    ...:         system_id = item["systemId"]
    ...:         current_id = current_id + (item_kind, system_id)
    ...:         data[current_id] = item
    ...:     for kind in item_kinds:
    ...:         if kind in item:
    ...:             kind_item = item[kind]
    ...:             data = get_item_data(kind_item, kind, current_id, item_kinds, data)
    ...:     return data
    ...: 

In [32]: data = get_item_data(fixture, "Backyard", (), item_kinds, data)

In [33]: len(data)
Out[33]: 24

In [34]: list(data)
Out[34]: 
[('Backyard', '24353'),
 ('Backyard', '24353', 'BOWS', '1'),
 ('Backyard', '24353', 'BOWS', '1', 'Filter', '2'),
 ('Backyard', '24353', 'BOWS', '1', 'VirtualHeater', '15'),
 ('Backyard', '24353', 'BOWS', '1', 'Heater', '16'),
 ('Backyard', '24353', 'BOWS', '1', 'CSAD', '0'),
 ('Backyard', '24353', 'BOWS', '1', 'Group', '34'),
 ('Backyard', '24353', 'BOWS', '1', 'Lights', '11'),
 ('Backyard', '24353', 'BOWS', '1', 'Lights', '25'),
 ('Backyard', '24353', 'BOWS', '1', 'Relays', '5'),
 ('Backyard', '24353', 'BOWS', '1', 'Relays', '14'),
 ('Backyard', '24353', 'BOWS', '1', 'Relays', '26'),
 ('Backyard', '24353', 'Relays', '17'),
 ('Backyard', '36211'),
 ('Backyard', '36211', 'BOWS', '2'),
 ('Backyard', '36211', 'BOWS', '2', 'Filter', '3'),
 ('Backyard', '36211', 'BOWS', '2', 'VirtualHeater', '4'),
 ('Backyard', '36211', 'BOWS', '2', 'Heater', '5'),
 ('Backyard', '36211', 'BOWS', '2', 'CSAD', '0'),
 ('Backyard', '36211', 'BOWS', '2', 'Lights', '23'),
 ('Backyard', '36211', 'BOWS', '2', 'Lights', '24'),
 ('Backyard', '36211', 'BOWS', '2', 'Relays', '7'),
 ('Backyard', '36211', 'Relays', '14'),
 ('Backyard', '36211', 'Relays', '29')]

In [35]: data['Backyard', '24353', 'BOWS', '1', 'Filter', '2']
Out[35]: 
{'systemId': '2',
 'valvePosition': '1',
 'filterSpeed': '90',
 'filterState': '1',
 'lastSpeed': '50',
 'Name': 'Filter Pump',
 'Shared-Type': 'BOW_NO_EQUIPMENT_SHARED',
 'Filter-Type': 'FMT_VARIABLE_SPEED_PUMP',
 'Max-Pump-Speed': '100',
 'Min-Pump-Speed': '18',
 'Max-Pump-RPM': '3450',
 'Min-Pump-RPM': '600',
 'Priming-Enabled': 'yes',
 'Alarms': []}
```